src/full_client.toit

```diff
@@ -221,7 +221,95 @@ interface ReconnectionStrategy:
 A base class for reconnection strategies.
 */
 abstract class DefaultReconnectionStrategyBase implements ReconnectionStrategy:
-  UNIMPLEMENTED ::= "UNIMPLEMENTED"
+  static DEFAULT_RECEIVE_CONNECT_TIMEOUT /Duration ::= Duration --s=5
+  static DEFAULT_ATTEMPT_DELAYS /List ::= [
+    Duration --s=1,
+    Duration --s=5,
+    Duration --s=15,
+  ]
+
+  receive_connect_timeout_ /Duration
+  attempt_delays_ /List
+
+  is_closed_ := false
+
+  /**
+  A latch that is set when the client is sleeping before reconnecting.
+  */
+  waiting_for_reconnect_latch_ /monitor.Latch? := null
+
+  constructor
+      --receive_connect_timeout /Duration = DEFAULT_RECEIVE_CONNECT_TIMEOUT
+      --attempt_delays /List /*Duration*/ = DEFAULT_ATTEMPT_DELAYS:
+    receive_connect_timeout_ = receive_connect_timeout
+    attempt_delays_ = attempt_delays
+
+
+  /**
+  Tries to connect, potentially retrying with delays.
+
+  Returns null if the strategy was closed.
+  Returns whether the broker had a session for this client, otherwise.
+  */
+  do_connect transport/ActivityMonitoringTransport
+      [--reconnect_transport]
+      [--send_connect]
+      [--receive_connect_ack]:
+    for i := -1; i < attempt_delays_.size; i++:
+      if is_closed: return null
+      if i >= 0:
+        // TODO(florian): we want to wait for `close` calls and `sleep` at the same
+        // time. Using a task and latch for this feels very expensive.
+        assert: waiting_for_reconnect_latch_ == null
+        waiting_for_reconnect_latch_ = monitor.Latch
+        sleeping_task := null
+        try:
+          sleeping_task = task::
+            sleep attempt_delays_[i]
+            sleeping_task = null
+            if waiting_for_reconnect_latch_ and not waiting_for_reconnect_latch_.has_value:
+              waiting_for_reconnect_latch_.set true
+
+          // As soon as the sleep is over, or there is a close, we get here.
+          waiting_for_reconnect_latch_.get
```

Comment:
Can't you just use a timeout instead of having a sleep in a separate task?

```suggestion
          catch --unwind=(: it != DEADLINE_EXCEEDED_ERROR):
            with_timeout attempt_delays[i]: waiting_for_reconnect_latch_.get
```

Reply:
Will do this in a follow-up PR. (as I can't run the code in this PR).

Reply:
https://github.com/toitware/mqtt/pull/32